File: src/simulation/marching-cubes.cpp
#include "marching-cubes.h"

#include <vector>

#include "SimpleMath.h"

using namespace std;

void MarchingCube::update_grid() {
  for (int i = 0; i < m_voxel_grid.data.size(); i++) {
    m_voxel_grid.data[i] = false;
  }
  for (auto& p : m_particles) {
    XMINT3 dPos = XMINT3(std::round(p.position.x / m_width),
                         std::round(p.position.y / m_width),
                         std::round(p.position.z / m_width));
    int radius = std::ceil(m_particle_radius / m_width);
    for (int i = -radius; i <= radius; i++) {
      for (int j = -radius; j <= radius; j++) {
        for (int k = -radius; k <= radius; k++) {
          XMINT3 cubePos(i + dPos.x, j + dPos.y, k + dPos.z);
          if (i + dPos.x < 0 || j + dPos.y < 0 || k + dPos.z < 0 ||
              i + dPos.x >= m_num.x || j + dPos.y >= m_num.y ||
              k + dPos.z >= m_num.z) {
            continue;
          }

          int index = cubePos.x + (cubePos.z * m_num.y + cubePos.y) * m_num.x;
          if (!m_voxel_grid.data[index]) {
            m_voxel_grid.data[index] = check_collision(
                Vector3(cubePos.x, cubePos.y, cubePos.z) * m_width +
                    m_worldOffset,
                p.position);
          }
        }
      }
    }
  }

  // for (int i = 0; i < m_num.x; ++i) {
  //   for (int j = 0; j < m_num.y; ++j) {
  //     for (int k = 0; k < m_num.z; ++k) {
  //       int index = i + (k * m_num.y + j) * m_num.x;
  //       m_voxel_grid.data[index] =
  //           check_collision(Vector3(i, j, k) * m_width + m_worldOffset);
  //     }
  //   }
  // }
}

void MarchingCube::march(std::vector<Vector3>& vertex) {
  vertex.clear();
  update_grid();

  for (int y = 0; y < m_num.y - 1; y++) {
    for (int x = 0; x < m_num.x - 1; x++) {
      for (int z = 0; z < m_num.z - 1; z++) {
        // if (m_voxel_grid.data[x + (y + z * m_num.y) * m_num.x]) {
        //   vertex.push_back(Vector3(x, y, z) * m_width + m_worldOffset);
        // }
        march_cube(XMINT3(x, y, z), vertex);
      }
    }
  }
}

void MarchingCube::march_cube(XMINT3 pos, std::vector<Vector3>& vertex) {
  auto edges = get_triangulations(pos.x, pos.y, pos.z);
  for (int i = 0; i <= 15 && edges[i] != -1; i += 3) {
    Vector3 p1 = get_point(edges[i], pos);
    Vector3 p2 = get_point(edges[i + 1], pos);
    Vector3 p3 = get_point(edges[i + 2], pos);

    vertex.push_back(p1);
    vertex.push_back(p2);
    vertex.push_back(p3);
  }
}

Vector3 MarchingCube::get_point(UINT edge_index, XMINT3 pos) {
  auto point_indecies = EDGES_TABLE[edge_index];
  XMINT3 p1 = POINTS_TABLE[point_indecies[0]];
  XMINT3 p2 = POINTS_TABLE[point_indecies[1]];

  Vector3 worldP1 =
      (Vector3(p1.x, p1.y, p1.z) + Vector3(pos.x, pos.y, pos.z)) * m_width +
      m_worldOffset;
  Vector3 worldP2 =
      (Vector3(p2.x, p2.y, p2.z) + Vector3(pos.x, pos.y, pos.z)) * m_width +
      m_worldOffset;

  return (worldP1 + worldP2) / 2;
}

const int* MarchingCube::get_triangulations(UINT x, UINT y, UINT z) {
  /*

  indecies:
           (5)-------(6)
         .  |      .  |
      (4)-------(7)   |
       |    |    |    |
       |   (1)---|---(2)
       | .       | .
      (0)-------(3)

  */

  UINT idx = 0;
  idx |= (UINT)(!m_voxel_grid.get(x, y, z)) << 0;
  idx |= (UINT)(!m_voxel_grid.get(x, y, z + 1)) << 1;
  idx |= (UINT)(!m_voxel_grid.get(x + 1, y, z + 1)) << 2;
  idx |= (UINT)(!m_voxel_grid.get(x + 1, y, z)) << 3;
  idx |= (UINT)(!m_voxel_grid.get(x, y + 1, z)) << 4;
  idx |= (UINT)(!m_voxel_grid.get(x, y + 1, z + 1)) << 5;
  idx |= (UINT)(!m_voxel_grid.get(x + 1, y + 1, z + 1)) << 6;
  idx |= (UINT)(!m_voxel_grid.get(x + 1, y + 1, z)) << 7;

  assert(idx < 256);

  return TRIANGULATIONS[idx];
}

bool MarchingCube::check_collision(Vector3 point) {
  float min_dist = m_particle_radius;
  for (auto& p : m_particles) {
    float d = Vector3::Distance(point, p.position);
    if (d <= min_dist) {
      return true;
    }
  }
  return false;
}

bool MarchingCube::check_collision(const Vector3& point,
                                   const Vector3& particle) {
  float min_dist = m_particle_radius;
  float d = Vector3::Distance(point, particle);
  if (d <= min_dist) {
    return true;
  }
  return false;
}

File: shaders/marching-cubes/MarchingCubes.cs
#include "shaders/Sph.h"
#include "shaders/LookUpTable.h"


struct Triangulation
{
    int data[15];
};

struct Edges
{
    int data[2];
};

struct Triangle
{
    float3 pos[3];
    float3 norm[3];
};

StructuredBuffer<Particle> particles : register(t0);
RWStructuredBuffer<float> voxel_grid : register(u0);
AppendStructuredBuffer<Triangle> triangles : register(u1);
RWStructuredBuffer<SurfaceBuffer> surfaceBuffer : register(u2);

static const float g_c = 0.25;

float voxel_get(in uint x, in uint y, in uint z)
{
    return voxel_grid[x + y * MC_DIMENSIONS.x + z * MC_DIMENSIONS.x * MC_DIMENSIONS.y];
}



Triangulation get_triangulations(in uint x, in uint y, in uint z)
{
  /*

  indecies:
           (5)-------(6)
         .  |      .  |
      (4)-------(7)   |
       |    |    |    |
       |   (1)---|---(2)
       | .       | .
      (0)-------(3)

  */

    uint idx = 0;
    idx |= (uint) !(voxel_get(x, y, z) > g_c) << 0;
    idx |= (uint) !(voxel_get(x, y, z + 1) > g_c) << 1;
    idx |= (uint) !(voxel_get(x + 1, y, z + 1) > g_c) << 2;
    idx |= (uint) !(voxel_get(x + 1, y, z) > g_c) << 3;
    idx |= (uint) !(voxel_get(x, y + 1, z) > g_c) << 4;
    idx |= (uint) !(voxel_get(x, y + 1, z + 1) > g_c) << 5;
    idx |= (uint) !(voxel_get(x + 1, y + 1, z + 1) > g_c) << 6;
    idx |= (uint) !(voxel_get(x + 1, y + 1, z) > g_c) << 7;

    return TRIANGULATIONS[idx];
}

float3 get_gradient(in uint3 pos) {
  float3 gradient;

  float xp, xn, yp, yn, zp, zn;
  if (pos.x + 1 < MC_DIMENSIONS.x) {
    xp = voxel_grid[pos.x + 1 + pos.y * MC_DIMENSIONS.x + pos.z *
      MC_DIMENSIONS.x * MC_DIMENSIONS.y];
  }

  if (pos.y + 1 < MC_DIMENSIONS.y) {
    yp = voxel_grid[pos.x + (pos.y + 1) * MC_DIMENSIONS.x + pos.z *
      MC_DIMENSIONS.x * MC_DIMENSIONS.y];
  }

  if (pos.z + 1 < MC_DIMENSIONS.z) {
    zp = voxel_grid[pos.x + pos.y * MC_DIMENSIONS.x + (pos.z + 1) *
      MC_DIMENSIONS.x * MC_DIMENSIONS.y];
  }

  if (pos.x - 1 >= 0) {
    xn = voxel_grid[pos.x - 1 + pos.y * MC_DIMENSIONS.x + pos.z *
      MC_DIMENSIONS.x * MC_DIMENSIONS.y];
  }

  if (pos.y - 1 >= 0) {
    yn = voxel_grid[pos.x + (pos.y - 1) * MC_DIMENSIONS.x + pos.z *
      MC_DIMENSIONS.x * MC_DIMENSIONS.y];
  }

  if (pos.x - 1 >= 0) {
    zn = voxel_grid[pos.x + pos.y * MC_DIMENSIONS.x + (pos.z - 1) *
      MC_DIMENSIONS.x * MC_DIMENSIONS.y];
  }

  gradient.x = (xp - xn) / 2;
  gradient.y = (yp - yn) / 2;
  gradient.z = (zp - zn) / 2;
 
  return gradient;
}

struct Vertex {
  float3 pos;
  float3 norm;
};

Vertex get_point(in uint edge_index, in uint3 pos)
{
    Edges point_indecies = EDGES_TABLE[edge_index];
    uint3 p1 = POINTS_TABLE[point_indecies.data[0]];
    uint3 p2 = POINTS_TABLE[point_indecies.data[1]];

    float3 worldP1 =
        (p1 + pos + float3(0.5f, 0.5f, 0.5f)) * marchingWidth + worldPos;
    float3 worldP2 =
        (p2 + pos + float3(0.5f, 0.5f, 0.5f)) * marchingWidth + worldPos;

    uint3 cell1 = p1 + pos;
    uint index1 = cell1.x + cell1.y * MC_DIMENSIONS.x + cell1.z * MC_DIMENSIONS.x * MC_DIMENSIONS.y;
    uint3 cell2 = p2 + pos;
    uint index2 = cell2.x + cell2.y * MC_DIMENSIONS.x + cell2.z * MC_DIMENSIONS.x * MC_DIMENSIONS.y;
    float alpha = (g_c - voxel_grid[index1]) / (voxel_grid[index2] - voxel_grid[index1]);


    Vertex res;

    res.pos = (1 - alpha) * worldP1 + alpha * worldP2;

    float t = (g_c - voxel_grid[index1]) / (voxel_grid[index2] - voxel_grid[index1]);
    float3 normal1 = get_gradient(cell1);
    float3 normal2 = get_gradient(cell2);

    res.norm = normalize(lerp(normal1, normal2, t));

    return res;
}

void march_cube(in uint3 pos)
{
    Triangulation edges = get_triangulations(pos.x, pos.y, pos.z);
    for (int i = 0; i < 15 && edges.data[i] != -1; i += 3)
    {
        Triangle res;
        Vertex v;
        v = get_point(edges.data[i], pos);
        res.pos[0] = v.pos;
        res.norm[0] = v.norm;
        v = get_point(edges.data[i + 1], pos);
        res.pos[1] = v.pos;
        res.norm[1] = v.norm;
        v = get_point(edges.data[i + 2], pos);
        res.pos[2] = v.pos;
        res.norm[2] = v.norm;
        triangles.Append(res);
    }
}


void march(in uint index)
{
    uint3 cell = GetMCCell(index);

    if (cell.x >= MC_DIMENSIONS.x - 1 || cell.y >= MC_DIMENSIONS.y - 1 || cell.z >= MC_DIMENSIONS.z - 1)
    {
      return;
    }

    march_cube(cell);
}

[numthreads(BLOCK_SIZE, 1, 1)]
void cs(uint3 globalThreadId : SV_DispatchThreadID)
{
    march(globalThreadId.x);
}
